server/schedulers/evict_slow_store.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2021 TiKV Project Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package schedulers
+
+import (
+	"strconv"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/pingcap/log"
+	"github.com/tikv/pd/pkg/errs"
+	"github.com/tikv/pd/server/core"
+	"github.com/tikv/pd/server/schedule"
+	"github.com/tikv/pd/server/schedule/operator"
+	"github.com/tikv/pd/server/schedule/opt"
+	"go.uber.org/zap"
+)
+
+const (
+	// EvictSlowStoreName is evict leader scheduler name.
+	EvictSlowStoreName = "evict-slow-store-scheduler"
+	// EvictSlowStoreType is evict leader scheduler type.
+	EvictSlowStoreType = "evict-slow-store"
+
+	slowStoreEvictThreshold   = 100
+	slowStoreRecoverThreshold = 1
+)
+
+func init() {
+	schedule.RegisterSliceDecoderBuilder(EvictSlowStoreType, func(args []string) schedule.ConfigDecoder {
+		return func(v interface{}) error {
+			if len(args) != 1 && len(args) != 0 {
+				return errs.ErrSchedulerConfig.FastGenByArgs("evicted-store")
+			}
+			conf, ok := v.(*evictSlowStoreSchedulerConfig)
+			if !ok {
+				return errs.ErrScheduleConfigNotExist.FastGenByArgs()
+			}
+			if len(args) == 1 {
+				id, err := strconv.ParseUint(args[0], 10, 64)
+				if err != nil {
+					return errs.ErrStrconvParseUint.Wrap(err).FastGenWithCause()
+				}
+				conf.EvictedStores = []uint64{id}
+			}
+			return nil
+		}
+	})
+
+	schedule.RegisterScheduler(EvictSlowStoreType, func(opController *schedule.OperatorController, storage *core.Storage, decoder schedule.ConfigDecoder) (schedule.Scheduler, error) {
+		conf := &evictSlowStoreSchedulerConfig{storage: storage, EvictedStores: make([]uint64, 0)}
+		if err := decoder(conf); err != nil {
+			return nil, err
+		}
+		return newEvictSlowStoreScheduler(opController, conf), nil
+	})
+}
+
+type evictSlowStoreSchedulerConfig struct {
+	storage       *core.Storage
+	EvictedStores []uint64 `json:"evict-stores"`
+}
+
+func (conf *evictSlowStoreSchedulerConfig) Persist() error {
+	name := conf.getSchedulerName()
+	data, err := schedule.EncodeConfig(conf)
+	failpoint.Inject("persistFail", func() {
+		err = errors.New("fail to persist")
+	})
+	if err != nil {
+		return err
+	}
+	return conf.storage.SaveScheduleConfig(name, data)
+}
+
+func (conf *evictSlowStoreSchedulerConfig) getSchedulerName() string {
+	return EvictSlowStoreName
+}
+
+type evictSlowStoreScheduler struct {
+	*BaseScheduler
+	conf                 *evictSlowStoreSchedulerConfig
+	evictLeaderScheduler *evictLeaderScheduler
```

Comment:
I think it is not a very good idea to use another scheduler as a toolkit because each scheduler has its own context. For example, scheduler names will be recorded to the filter metrics, it may happen that we want to record evict-slow-store-scheduler and end up with evict-leader-scheduler.
There are two possible ways to improve this: a) reimplement the similar code again b) extract the common parts and let different schedulers call them

Reply:
The logic is basically the same so I think it would be better to reuse it directly. About the metrics, I saw that it is based on the name of the scheduler, may setting a customized name is a good solution?

Reply:
Consider extracting the common part?